include/gatekeeper_fib.h

```diff
@@ -0,0 +1,202 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_GK_FIB_H_
+#define _GATEKEEPER_GK_FIB_H_
+
+#include <rte_ether.h>
+#include <rte_hash.h>
+
+#include "gatekeeper_net.h"
+#include "gatekeeper_lpm.h"
+
+enum gk_fib_action {
+
+	/* Forward the packet to the corresponding Grantor. */
+	GK_FWD_GRANTOR,
+
+	/*
+	 * Forward the packet to the corresponding gateway
+	 * in the front network.
+	 */
+	GK_FWD_GATEWAY_FRONT_NET,
+
+	/*
+	 * Forward the packet to the corresponding gateway
+	 * in the back network.
+	 */
+	GK_FWD_GATEWAY_BACK_NET,
+
+	/*
+	 * The destination address is a neighbor in the front network.
+	 * Forward the packet to the destination directly.
+	 */
+	GK_FWD_NEIGHBOR_FRONT_NET,
+
+	/*
+	 * The destination address is a neighbor in the back network.
+	 * Forward the packet to the destination directly.
+	 */
+	GK_FWD_NEIGHBOR_BACK_NET,
+
+	/* Drop the packet. */
+	GK_DROP,
+
+	/* Invalid forward action. */
+	GK_FIB_MAX,
+};
+
+/* The Ethernet header cache. */
+struct ether_cache {
+
+	/* Indicate whether the cache entry has been filled in or not. */
+	bool             filled;
+
+	/* Indicate whether the MAC address is stale or not. */
+	bool             stale;
+
+	/* The whole Ethernet header. */
+	struct ether_hdr eth_hdr;
+
+	/*
+	 * The count of how many times the LPM tables refer to it,
+	 * so a neighbor entry can go away only when no one referring to it.
+	 */
+	uint32_t         ref_cnt;
+};
+
+struct neighbor_hash_table {
+
+	/* The tables that store the Ethernet headers. */
+	struct ether_cache *cache_tbl;
+
+	struct rte_hash *hash_table;
+};
+
+/* The gk forward information base (fib). */
+struct gk_fib {
+
+	/* The fib action. */
+	enum gk_fib_action action;
+
+	/*
+	 * The caller that finished updating its flow table
+	 * needs to increment this counter, so that the blocks
+	 * that updating the FIB entry can finish its operation.
+	 */
+	rte_atomic16_t     num_updated_instances;
```

Comment:
Improve the comment to account for the fact that GK blocks will increment this variable not only to flush their flow tables.